### ISSUE TITLE
web-tree-sitter query with predicates fills array with undefined values

### ISSUE THREAD

**Issue by cellog** [tree-sitter/tree-sitter#932]
Simple example to run in node:

```js
const Parser = require("web-tree-sitter");
const path = require("path");
Parser.init().then(async () => {
  const parser = new Parser();
  const Javascript = await Parser.Language.load(
    path.resolve(__dirname, "tree-sitter-javascript.wasm")
  );
  parser.setLanguage(Javascript);
  const query = Javascript.query(`(
    call_expression
      function: (
        member_expression
          object: (identifier) @I18n
                  (#eq? @I18n "I18n")
          property: (property_identifier) @ns
                    (#eq? @ns "namespace")
      )
      arguments: (
        arguments (string) @namespace
      )
  )`);
  const tree = parser.parse(`var p, t;

  ({t} = I18n.namespace("blah"));
  
  ({p} = Some.other("thing"));
  `);
  console.log(query.matches(tree.rootNode));
})
```

output is:

```bash
[
  { pattern: 0, captures: [ [Object], [Object], [Object] ] },
  <1 empty item>
]
```

Looking at https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/binding.js#L975 the reason for this is pretty obvious. This code (https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/binding.js#L972) assumes the array will contain all of the captures:

```js
const result = new Array(count);
```

but this code (https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/binding.js#L983)

```js
      if (this.textPredicates[pattern].every(p => p(captures))) {
        result[i] = {pattern, captures};
```

only runs if all of the `textPredicates` match that capture - and it sets the current capture index. In the example above, the first instance (` ({t} = I18n.namespace("blah"));`) matches, which is capture index 0. The second instance does not match, and since the array has been pre-allocated with 2 entries, it results in a 2nd entry of `undefined`. `query.captures()` does not suffer from this bug.

The solution is not to pre-allocate the array length, but to ~~either use `result.push()` instead of `result[i] = ` or if this is not allowed (not sure how restrictive js -> wasm is), do 2 passes, the first to determine which predicates match, and then allocate the array and fill it.~~ use `result.push()` the way that `captures` does.

**Comment by maxbrunsfeld** [comment]
Thank you!

**Comment by maxbrunsfeld** [comment]
This is fixed in `web-tree-sitter` v0.18.2.
